From 02bc7e1437fb2d846de7d958ee4b9c8b28c211c9 Mon Sep 17 00:00:00 2001
From: huawei <huawei@huawei.com>
Date: Mon, 3 Apr 2023 20:15:10 +0800
Subject: [PATCH 2/2] ascend: add emmc patch for ascend 310B

Offering:ascend
Reference:NA
Type:requirement
DTS/AR:NA
reason:emmc mbigen
---
 drivers/mmc/host/Makefile           |   3 +-
 drivers/mmc/host/mci_proc.c         | 343 ++++++++++++++++++++++++++++
 drivers/mmc/host/mci_proc.h         |  18 ++
 drivers/mmc/host/sdhci.c            |  10 +
 drivers/mmc/host/sdhci_zodiac_mmc.c |   4 +
 5 files changed, 377 insertions(+), 1 deletion(-)
 create mode 100644 drivers/mmc/host/mci_proc.c
 create mode 100644 drivers/mmc/host/mci_proc.h

diff --git a/drivers/mmc/host/Makefile b/drivers/mmc/host/Makefile
index 53590de73..7aadf0f41 100644
--- a/drivers/mmc/host/Makefile
+++ b/drivers/mmc/host/Makefile
@@ -16,7 +16,7 @@ sdhci-pci-y			+= sdhci-pci-core.o sdhci-pci-o2micro.o sdhci-pci-arasan.o \
 				   sdhci-pci-dwc-mshc.o sdhci-pci-gli.o
 obj-$(subst m,y,$(CONFIG_MMC_SDHCI_PCI))	+= sdhci-pci-data.o
 obj-$(CONFIG_MMC_SDHCI_ZODIAC_310B)      += sdhci_zodiac_mmc_310b.o
-sdhci_zodiac_mmc_310b-objs := sdhci_zodiac_mmc.o mmc_misc_1910b.o emmc_phy.o
+sdhci_zodiac_mmc_310b-objs := sdhci_zodiac_mmc.o mmc_misc_1910b.o emmc_phy.o mci_proc.o
 obj-$(CONFIG_MMC_SDHCI_ACPI)	+= sdhci-acpi.o
 obj-$(CONFIG_MMC_SDHCI_PXAV3)	+= sdhci-pxav3.o
 obj-$(CONFIG_MMC_SDHCI_PXAV2)	+= sdhci-pxav2.o
@@ -114,3 +114,4 @@ endif
 
 obj-$(CONFIG_MMC_SDHCI_XENON)	+= sdhci-xenon-driver.o
 sdhci-xenon-driver-y		+= sdhci-xenon.o sdhci-xenon-phy.o
+
diff --git a/drivers/mmc/host/mci_proc.c b/drivers/mmc/host/mci_proc.c
new file mode 100644
index 000000000..71133c7ff
--- /dev/null
+++ b/drivers/mmc/host/mci_proc.c
@@ -0,0 +1,343 @@
+/*****************************************************************************
+ *  This is the driver for the host mci SOC.
+ *
+ *  Copyright (C) Hisilicon. All rights reserved.
+ *
+ ******************************************************************************/
+
+#include <linux/proc_fs.h>
+#include <linux/seq_file.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <linux/platform_device.h>
+#include <linux/mmc/card.h>
+#include <linux/mmc/core.h>
+#include <linux/mmc/host.h>
+#include "sdhci.h"
+#include "mci_proc.h"
+#include "../core/card.h"
+#include "../core/core.h"
+
+#define MCI_PARENT       "mci"
+#define MCI_STATS_PROC   "mci_info"
+#define MAX_CLOCK_SCALE	(4)
+#define UNSTUFF_BITS(resp,start,size)                   \
+	({                              \
+	 const int __size = size;                \
+	 const u32 __mask = (__size < 32 ? 1 << __size : 0) - 1; \
+	 const int __off = 3 - ((start) / 32);           \
+	 const int __shft = (start) & 31;            \
+	 u32 __res;                      \
+	 \
+	 __res = resp[__off] >> __shft;              \
+	 if (__size + __shft > 32)               \
+	 __res |= resp[__off-1] << ((32 - __shft) % 32); \
+	 __res & __mask;                     \
+	 })
+
+extern unsigned long long g_write_blkcount;
+
+struct card_info {
+	unsigned int type;
+	unsigned int state;
+	unsigned int sd_bus_speed;
+	unsigned int raw_ssr[16];
+};
+
+unsigned int slot_index = 0;
+struct sdhci_host *mci_host[MCI_SLOT_NUM] = {NULL};
+static struct proc_dir_entry *proc_mci_dir;
+
+static char *card_type[MAX_CARD_TYPE + 1] = {
+	"MMC card",
+	"SD card",
+	"SDIO card",
+	"SD combo (IO+mem) card",
+	"unknown"
+};
+static char *clock_unit[4] = {
+	"Hz",
+	"KHz",
+	"MHz",
+	"GHz"
+};
+
+static char *mci_get_card_type(unsigned int sd_type)
+{
+	if (MAX_CARD_TYPE <= sd_type)
+		return card_type[MAX_CARD_TYPE];
+	else
+		return card_type[sd_type];
+}
+
+static unsigned int analyze_clock_scale(unsigned int clock,
+		unsigned int *clock_val)
+{
+	unsigned int scale = 0;
+	unsigned int tmp = clock;
+
+	while (1) {
+		tmp = tmp / 1000;
+		if (0 < tmp) {
+			*clock_val = tmp;
+			scale++;
+		} else {
+			break;
+		}
+	}
+	return scale;
+}
+
+static inline int mmc_in_uhs(struct mmc_host *host)
+{
+	return host->ios.timing >= MMC_TIMING_UHS_SDR12 &&
+		host->ios.timing <= MMC_TIMING_UHS_DDR50;
+};
+
+static inline int mmc_in_hs(struct mmc_host *host)
+{
+	return host->ios.timing == MMC_TIMING_SD_HS ||
+		host->ios.timing == MMC_TIMING_MMC_HS;
+};
+
+static inline int mmc_in_hs400(struct mmc_host *host)
+{
+	return host->ios.timing == MMC_TIMING_MMC_HS400;
+};
+
+static inline int mmc_in_hs200(struct mmc_host *host)
+{
+	return host->ios.timing == MMC_TIMING_MMC_HS200;
+};
+
+static inline int mmc_in_ddr52(struct mmc_host *host)
+{
+	return host->ios.timing == MMC_TIMING_MMC_DDR52;
+};
+
+static void mci_stats_seq_printout(struct seq_file *s)
+{
+	unsigned int index_mci;
+	unsigned int clock;
+	unsigned int clock_scale;
+	unsigned int clock_value = 0;
+	const char *type;
+	static struct mmc_host *mmc = NULL;
+	static struct mmc_card	*card;
+	const char *uhs_bus_speed_mode = "";
+	static const char *const uhs_speeds[] = {
+		[UHS_SDR12_BUS_SPEED] = "SDR12 ",
+		[UHS_SDR25_BUS_SPEED] = "SDR25 ",
+		[UHS_SDR50_BUS_SPEED] = "SDR50 ",
+		[UHS_SDR104_BUS_SPEED] = "SDR104 ",
+		[UHS_DDR50_BUS_SPEED] = "DDR50 ",
+	};
+	unsigned int speed_class, grade_speed_uhs;
+	struct card_info info = {0};
+	unsigned int present;
+
+	for (index_mci = 0; index_mci < MCI_SLOT_NUM; index_mci++) {
+		if (NULL == mci_host[index_mci]) {
+			seq_printf(s, "MCI%d invalid\n", index_mci);
+			continue;
+		} else {
+			mmc = mci_host[index_mci]->mmc;
+			seq_printf(s, "MCI%d", index_mci);
+		}
+
+		present = mmc->ops->get_cd(mmc);
+		if (present) {
+			seq_puts(s, ": pluged");
+		} else {
+			seq_puts(s, ": unplugged");
+		}
+
+		card = mmc->card;
+		if (NULL == card) {
+			seq_puts(s, "_disconnected\n");
+		} else {
+			mmc_get_card(card, NULL);
+			if (mmc_card_removed(card)) {
+				mmc_put_card(card, NULL);
+				seq_puts(s, "_disconnected\n");
+			}
+			info.type = card->type;
+			info.state = card->state;
+			info.sd_bus_speed = card->sd_bus_speed;
+			memcpy(info.raw_ssr, card->raw_ssr, 64);
+			mmc_put_card(card, NULL);
+
+			seq_puts(s, "_connected\n");
+
+			seq_printf(s,
+					"\tType: %s",
+					mci_get_card_type(info.type)
+					);
+
+			if (mmc_card_blockaddr(&info)) {
+				if (mmc_card_ext_capacity(&info))
+					type = "SDXC";
+				else
+					type = "SDHC";
+				seq_printf(s, "(%s)\n", type);
+			}
+
+			if(! strncmp(mci_get_card_type(info.type), "MMC card", 8))
+
+			{
+				seq_printf(s, "\tManfid:0x%02x\n", card->cid.manfid);
+				seq_printf(s, "\tBlkcount:0x%llx\n", g_write_blkcount);
+			}
+
+			if (mmc_in_uhs(mmc) &&
+					(info.sd_bus_speed < ARRAY_SIZE(uhs_speeds)))
+				uhs_bus_speed_mode = uhs_speeds[info.sd_bus_speed];
+			else
+				uhs_bus_speed_mode = "";
+
+			seq_printf(s, "\tMode: %s%s%s%s\n",
+					mmc_in_uhs(mmc) ? "UHS " :
+					(mmc_in_hs(mmc) ? "HS " : ""),
+					mmc_in_hs400(mmc) ? "HS400 " :
+					(mmc_in_hs200(mmc) ? "HS200 " : ""),
+					mmc_in_ddr52(mmc) ? "DDR " : "",
+					uhs_bus_speed_mode);
+
+			speed_class = UNSTUFF_BITS(info.raw_ssr, 440 - 384, 8);
+			grade_speed_uhs = UNSTUFF_BITS(info.raw_ssr, 396 - 384, 4);
+			seq_printf(s, "\tSpeed Class: Class %s\n",
+					(0x00 == speed_class) ? "0":
+					(0x01 == speed_class) ? "2":
+					(0x02 == speed_class) ? "4":
+					(0x03 == speed_class) ? "6":
+					(0x04 == speed_class) ? "10":
+					"Reserved");
+			seq_printf(s, "\tUhs Speed Grade: %s\n",
+					(0x00 == grade_speed_uhs)?
+					"Less than 10MB/sec(0h)" :
+					(0x01 == grade_speed_uhs)?
+					"10MB/sec and above(1h)":
+					"Reserved");
+
+			clock = mmc->ios.clock;
+			clock_scale = analyze_clock_scale(clock, &clock_value);
+			seq_printf(s, "\tHost work clock %d%s\n",
+					clock_value, clock_unit[clock_scale]);
+
+			clock = mmc->ios.clock;
+			clock_scale = analyze_clock_scale(clock, &clock_value);
+			seq_printf(s, "\tCard support clock %d%s\n",
+					clock_value, clock_unit[clock_scale]);
+			clock = mmc->actual_clock;
+			clock_scale = analyze_clock_scale(clock, &clock_value);
+			seq_printf(s, "\tCard work clock %d%s\n",
+					clock_value, clock_unit[clock_scale]);
+
+			//host = mmc_priv(mmc);
+			/* add card read/write error count */
+			//seq_printf(s, "\tCard error count: %d\n",
+			//		        host->error_count);
+		}
+	}
+}
+
+/* proc interface setup */
+static void *mci_seq_start(struct seq_file *s, loff_t *pos)
+{
+	/*   counter is used to tracking multi proc interfaces
+	 *  We have only one interface so return zero
+	 *  pointer to start the sequence.
+	 */
+	static unsigned long counter;
+
+	if (*pos == 0)
+		return &counter;
+
+	*pos = 0;
+	return NULL;
+}
+
+/* proc interface next */
+static void *mci_seq_next(struct seq_file *s, void *v, loff_t *pos)
+{
+	(*pos)++;
+	if (*pos >= MCI_SLOT_NUM)
+		return NULL;
+
+	return NULL;
+}
+
+/* define parameters where showed in proc file */
+static int mci_stats_seq_show(struct seq_file *s, void *v)
+{
+	mci_stats_seq_printout(s);
+	return 0;
+}
+
+/* proc interface stop */
+static void mci_seq_stop(struct seq_file *s, void *v)
+{
+}
+
+/* proc interface operation */
+static const struct seq_operations mci_stats_seq_ops = {
+	.start = mci_seq_start,
+	.next = mci_seq_next,
+	.stop = mci_seq_stop,
+	.show = mci_stats_seq_show
+};
+
+/* proc file open*/
+static int mci_stats_proc_open(struct inode *inode, struct file *file)
+{
+	return seq_open(file, &mci_stats_seq_ops);
+};
+
+/* proc file operation */
+static const struct proc_ops mci_stats_proc_ops = {
+  .proc_open = mci_stats_proc_open,
+  .proc_read = seq_read,
+  .proc_lseek = seq_lseek,
+  .proc_release = seq_release,
+};
+
+int mci_proc_init(struct sdhci_host *host)
+{
+	struct proc_dir_entry *proc_stats_entry;
+	int ret;
+
+	pr_info("entry mci proc init.current index :%d", slot_index);
+
+	if (slot_index == 0) {
+		proc_mci_dir = proc_mkdir(MCI_PARENT, NULL);
+		ret = PTR_ERR(proc_mci_dir);
+		if (!proc_mci_dir) {
+			pr_err("%s: failed to create proc file %s,ret = %d\n",
+					__func__, MCI_PARENT, ret);
+			return 1;
+		}
+
+		proc_stats_entry = proc_create(MCI_STATS_PROC,
+				0, proc_mci_dir, &mci_stats_proc_ops);
+		if (!proc_stats_entry) {
+			pr_err("%s: failed to create proc file %s\n",
+					__func__, MCI_STATS_PROC);
+			return 1;
+		}
+	}
+
+	mci_host[slot_index++] = host;
+
+	return 0;
+}
+
+int mci_proc_shutdown(void)
+{
+	if (proc_mci_dir) {
+		remove_proc_entry(MCI_STATS_PROC, proc_mci_dir);
+		remove_proc_entry(MCI_PARENT, NULL);
+		proc_mci_dir = NULL;
+	}
+
+	return 0;
+}
diff --git a/drivers/mmc/host/mci_proc.h b/drivers/mmc/host/mci_proc.h
new file mode 100644
index 000000000..b1ff97a46
--- /dev/null
+++ b/drivers/mmc/host/mci_proc.h
@@ -0,0 +1,18 @@
+/*
+ *  MCI connection table manager
+ */
+#ifndef __MCI_PROC_H__
+#define __MCI_PROC_H__
+
+#include <linux/proc_fs.h>
+
+#define MAX_CARD_TYPE	4
+#define MAX_SPEED_MODE	5
+
+#define MCI_SLOT_NUM 2
+extern unsigned int slot_index;
+
+int mci_proc_init(struct sdhci_host *host);
+int mci_proc_shutdown(void);
+
+#endif /*  __MCI_PROC_H__ */
diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 133f0d376..265086192 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -45,6 +45,10 @@
 
 #define MAX_TUNING_LOOP 40
 
+unsigned long long g_write_blkcount = 0;
+ 
+EXPORT_SYMBOL_GPL(g_write_blkcount);
+
 static unsigned int debug_quirks = 0;
 static unsigned int debug_quirks2;
 
@@ -1093,6 +1097,11 @@ static void sdhci_prepare_data(struct sdhci_host *host, struct mmc_command *cmd)
 {
 	struct mmc_data *data = cmd->data;
 
+	if ((host->mmc->card) && (MMC_TYPE_MMC == host->mmc->card->type) && (data->flags && MMC_DATA_WRITE ))
+	{
+		g_write_blkcount = g_write_blkcount + data->blocks;
+	}
+
 	sdhci_initialize_data(host, data);
 
 	if (host->flags & (SDHCI_USE_SDMA | SDHCI_USE_ADMA)) {
@@ -1181,6 +1190,7 @@ static void sdhci_prepare_data(struct sdhci_host *host, struct mmc_command *cmd)
 	sdhci_set_transfer_irqs(host);
 
 	sdhci_set_block_info(host, data);
+
 }
 
 #if IS_ENABLED(CONFIG_MMC_SDHCI_EXTERNAL_DMA)
diff --git a/drivers/mmc/host/sdhci_zodiac_mmc.c b/drivers/mmc/host/sdhci_zodiac_mmc.c
index 162136394..b1f800f51 100755
--- a/drivers/mmc/host/sdhci_zodiac_mmc.c
+++ b/drivers/mmc/host/sdhci_zodiac_mmc.c
@@ -24,6 +24,7 @@
 #include "sdhci_zodiac_mmc.h"
 #include "emmc_phy.h"
 #include "mmc_misc.h"
+#include "mci_proc.h"
 
 // controller new register
 #define SDHCI_CRG_CTRL0 0x80
@@ -873,6 +874,7 @@ static int sdhci_zodiac_probe(struct platform_device *pdev)
     struct sdhci_host *host = NULL;
     struct sdhci_zodiac_host *zodiac_host = NULL;
 
+
     dev_info(&pdev->dev, "sdhci_zodiac_probe start:\n");
     host = sdhci_pltfm_init(pdev, &sdhci_zodiac_pdata, sizeof(*zodiac_host));
     if (IS_ERR(host)) {
@@ -910,6 +912,8 @@ static int sdhci_zodiac_probe(struct platform_device *pdev)
         goto err_pltfm_free;
     }
 
+	ret = mci_proc_init(host);
+
     dev_info(&pdev->dev, "%s: sdhci_zodiac_probe done! \n", dev_name(zodiac_host->dev));
     return 0;
 
-- 
2.25.1

